SpringKonstant/src/org/usfirst/frc6357/robotcode/subsystems/TFMini.java

```diff
@@ -148,91 +160,80 @@ private void initialize()
+                                checksum = bytes[0];
+                                packet[count] = bytes[0];
+
+                                // Start looking for the next character which we hope is another 0x59.
+                                state = ReceivingState.WAIT_START2;
+                            }
+                            break;
+                        case WAIT_START2:
+                            if (bytes[0] == startByte) {
+                                // We received a second 0x59 so this really looks like a new packet.
+                                // Store the byte, update the checksum and byte counter then start reading
+                                // bytes until we reach the end of the packet.
+                                count++;
+                                checksum += bytes[0];
+                                packet[count] = bytes[0];
+                                state = ReceivingState.RECEIVING;
+                            } else {
+                                // This wasn't an 0x59 so we are not at the start of a packet. Go back and
+                                // start looking for a new packet header.
+                                state = ReceivingState.WAIT_START1;
+                            }
+                            break;
+                        case RECEIVING:
+                            // Save this byte in our packet buffer.
+                            count++;
+                            packet[count] = bytes[0];
+
+                            // Have we read a whole packet of data?
+                            if (count == PACKET_LENGTH - 1) {
+                                // Reached the end of the packet, so the byte we just received is the
+                                // checksum. Make sure it agrees with the checksum we've been accumulating
+                                // as we read packet bytes.
+                                if (checksum == bytes[0]) {
+                                    // Checksum is good so extract the distance, strength and quality info from
+                                    // the received packet.
+                                    parsePacket();
+                                }
+
+                                // Start looking for the start of the next packet.
+                                state = ReceivingState.WAIT_START1;
+                            } else {
+                                // We're not at the end of the packet so just update our checksum
+                                // accumulator and keep receiving.
+                                checksum += bytes[0];
+                            }
+                            break;
+                    }
+                }
+                Thread.sleep(10);
             }
-            catch(Exception e) {}
-        }).start();
+        } catch (Exception e) {
+        }
+
     }
 
     public void parsePacket()
     {
```

Comment:
Here, we use the synchronized(this) block again just to make sure our update of the variables isn't interrupted by a read of the variables, which could cause a bad value or values from different measurements to be read.